### ISSUE TITLE
Add support for pack application to docker container

### ISSUE THREAD

**Issue by negasus** [tarantool/cartridge-cli#18]
Add support for pack application to docker container, by analogy pack to rpm package

Expected command for example:
`cartridge pack docker --image docker-image-name:v0.0.0 ./app-dir`

**Comment by artur-barsegyan** [comment]
@dokshina What about your RFC? Let's discuss it here.

**Comment by dokshina** [comment]
@artur-barsegyan 

# RFC

## Pack application

`cartridge pack docker --version 1.0.0 ./myapp` provides docker image `myapp:1.0.0` with Cartridge application installed.

Application code (and Enterprise Tarantool binaries) is placed in `/opt/tarantool/myapp`.
Opensource Tarantool is installed on image.

Run directory is `/opt/tarantool/myapp/run`, workdir is `/opt/tarantool/myapp/workdir`.

## Start instance

```bash
$ docker run -d --name instance-1 myapp:1.0.0
```

Instance logs:

```bash
$ docker logs instance-1
```

## Configuration

Container env-file.

Example:

`env.instance-1`:

```bash
TARANTOOL_INSTANCE_NAME=instance-1
TARANTOOL_ADVERTISE_URI=3302
TARANTOOL_CLUSTER_COOKIE=secret
TARANTOOL_HTTP_PORT=8082
```

Start instance with this parameters.

```bash
$ docker run -d --env-file env.instance-1 \
                --name instance-1 \
                myapp:1.0.0
```

## Ports exposing

Ports exposing is a user responsibility.

For previous example:

```bash
$ docker run -d --env-file env.instance-1 \
                --name instance-1 \
                -p 127.0.0.1:8082:8082 \
                myapp:1.0.0
```

## Advertise URI

It's user responsibility to set up right advertise URI (`<host>:<port>`) if containers are deployed on different machines.

If user specifies only port, cartridge will use auto-detected IP, so user have to configure docker networks to set up instances communication.

## Connect to instance console

```bash
$ docker exec -ti instance-1 \
    tarantoolctl connect /opt/tarantool/myapp/run/myapp.control
```

## Update package

Now I see two ways:

* Use docker volumes to store instance `snap`s and `xlog`s on host machine (where exactly it must be stored on host?).

* Move instance files manually: stop old container, use `docker cp` to fetch workdir contents on host, then create new container, deliver instance files from the host using `docker cp` again and then start new container.

It seems like docker volumes way is much better.

## Ansible

All this actions can be pefrormed using Ansible.

**Comment by printercu** [comment]
Is it worth to use same advertise_uri and http port inside every container then map them to the real ports with `--publish`?

**Comment by artur-barsegyan** [comment]
> Opensource Tarantool is installed on image

What will be a base image?

**Comment by artur-barsegyan** [comment]
What about Lua -> Docker bindings? Is it works? Is it contains all that you need for current issue? Please, discover it and describe this too.

**Comment by dokshina** [comment]
> Is it worth to use same advertise_uri and http port inside every container then map them to the real ports with `--publish`?

I think it's OK for HTTP port, but it could be a problem for advertise_uri.
As I know, it must be equal to the URI that is used for asking instance outside of the container.

And I sure it would be better to use the same approach for both of this options.

**Comment by dokshina** [comment]
> What will be a base image?

`centos:7`

**Comment by knazarov** [comment]
Installation and runtime paths should be the same as on regular OS. /usr/share/tarantool, /var/lib/tarantool, /var/run/tarantool

**Comment by knazarov** [comment]
The base image should be centos:8

**Comment by knazarov** [comment]
> What about Lua -> Docker bindings? Is it works? Is it contains all that you need for current issue? Please, discover it and describe this too.

It is enough to just execute docker as a subprocess, in order to build images

**Comment by dokshina** [comment]
So, final RFC version:

@artur-barsegyan @knazarov 

# RFC

## Pack application

`cartridge pack docker --version 1.0.0 ./myapp` provides docker image `myapp:1.0.0` with Cartridge application installed.
The base image is `centos:8`.

Docker is executed as a subprocess to build image.

Application code (and Enterprise Tarantool binaries) is placed in `/usr/share/tarantool/myapp`.
Opensource Tarantool is installed on image.

Run directory is `/var/run/tarantool/myapp`, workdir is `/var/lib/tarantool/myapp`.

## Start instance

```bash
$ docker run -d --name instance-1 myapp:1.0.0
```

Instance logs:

```bash
$ docker logs instance-1
```

## Configuration

Container env-file.

Example:

`env.instance-1`:

```bash
TARANTOOL_INSTANCE_NAME=instance-1
TARANTOOL_ADVERTISE_URI=3302
TARANTOOL_CLUSTER_COOKIE=secret
TARANTOOL_HTTP_PORT=8082
```

Start instance with this parameters.

```bash
$ docker run -d --env-file env.instance-1 \
                --name instance-1 \
                myapp:1.0.0
```

## Ports exposing

Ports exposing is a user responsibility.

For previous example:

```bash
$ docker run -d --env-file env.instance-1 \
                --name instance-1 \
                -p 127.0.0.1:8082:8082 \
                myapp:1.0.0
```

## Advertise URI

It's user responsibility to set up right advertise URI (`<host>:<port>`) if containers are deployed on different machines.

If user specifies only port, cartridge will use auto-detected IP, so user have to configure docker networks to set up instances communication.

## Connect to instance console

```bash
$ docker exec -ti instance-1 \
    tarantoolctl connect /var/run/tarantool/myapp.control
```

## Update package

Docker volumes are used to store instance `snap`s and `xlog`s on host machine.
To start image with a new application code user needs to stop the old container and start a new one using new image.

## Ansible

All this actions can be pefrormed using Ansible.

**Comment by artur-barsegyan** [comment]
LGTM. We need to implement it @dokshina.

**Comment by dokshina** [comment]
@knazarov suggested building (running `tarantoolctl rocks make`) inside of container during `docker build`.
It looks OK for opensource Tarantool, but for enterprise Tarantool user needs to specify downoad credentials and version of SDK. 

Download credentials would be passed from env or by using `--download-token` argument for `cartridge pack docker` command.

To specify opensource Tarantool or SDK version `runtime.txt` file in project root will be used (like [here](https://devcenter.heroku.com/articles/python-runtimes)).

**Comment by artur-barsegyan** [comment]
For version specifier, we also can use a CLI argument.

**Comment by dokshina** [comment]
I think that using `runtime.txt` is better because it allows you to fix Tarantool version for your project.

**Comment by artur-barsegyan** [comment]
Yes, you right. It should be more convenient to have a project-specified file with this kind of setting. 

But you need to resolve some intersection with a `VERSION` file.

**Comment by dokshina** [comment]
Implemented in #75
